python/ray/llm/_internal/serve/deployments/data_parallel/dp_rank_assigner.py

```diff
@@ -12,15 +16,68 @@ class DPRankAssigner:
     deployment.
     """
 
-    def __init__(self, dp_size: int):
-        self.dp_size = dp_size
-        self.lock = asyncio.Lock()
-        self.next_rank = 0
-        self.dp_address = None
-        self.dp_rpc_port = None
-        self.master_info_event = asyncio.Event()
+    def __init__(self, dp_size: int, dp_size_per_node: Optional[int] = None):
+        self.dp_size: int = dp_size
+        self.dp_size_per_node: Optional[int] = dp_size_per_node
+        self.lock: asyncio.Lock = asyncio.Lock()
+        self.dp_address: Optional[str] = None
+        self.dp_rpc_port: Optional[int] = None
+        self.master_info_event: asyncio.Event = asyncio.Event()
+
+        # Fields for _register_random_placement():
+        # Next rank to assign
+        self.next_rank: Optional[int] = None
+
+        # Fields for _register_node_pack_placement():
+        # Number of nodes to assign to
+        self.num_nodes: Optional[int] = None
+        # Map from node id to available ranks
+        self.node_to_avail_ranks: Dict[str, List[int]] = {}
+
+        if dp_size_per_node is None:
+            self.next_rank = 0
+            logger.info(
+                f"Using random placement rank assigner for DP size {self.dp_size}"
+            )
+        else:
+            if self.dp_size_per_node <= 0:
+                raise ValueError(
+                    f"dp_size_per_node {self.dp_size_per_node} must be greater than 0"
+                )
+            if self.dp_size % self.dp_size_per_node != 0:
+                raise ValueError(
+                    f"dp_size {self.dp_size} must be divisible by dp_size_per_node {self.dp_size_per_node}"
+                )
+            self.num_nodes = self.dp_size // self.dp_size_per_node
+            logger.info(
+                f"Using node pack placement rank assigner for DP size {self.dp_size}"
+                f"with dp_size_per_node {self.dp_size_per_node}"
+            )
+
+    async def register(self, replica_ctx: "serve.context.ReplicaContext", node_id: str):
```

Comment:
node_id should be optional str if dp_size_per_node is set it should be passed in. o.w. the other code path won't need it.